File: task 2/hide.py
# coding: cp1251

# Массивы для русских и латинских букв-замен
rus_letters = ['а', 'е', 'о', 'р', 'с', 'у', 'х', 'А', 'В', 'Е', 'К', 'О', 'Р', 'С', 'Т', 'Х']
eng_letters = ['a', 'e', 'o', 'p', 'c', 'y', 'x', 'A', 'B', 'E', 'K', 'O', 'P', 'C', 'T', 'X']

# Чтение файлов
container_file = 'container.txt'
secret_message_file = 'secret.txt'
output_file = 'code.txt'

def text_to_bin(text):
    """Преобразование текста в бинарное представление с учетом кодировки CP1251."""
    binary_str = ''.join(format(ord(char), '08b') for char in text)
    return binary_str

def hide_info_in_container(container, secret_message_bin):
    """Скрытие информации в контейнере."""
    container_with_hidden_info = []
    message_idx = 0
    
    for char in container:
        if char in rus_letters and message_idx < len(secret_message_bin):
            if secret_message_bin[message_idx] == '1':
                # Заменяем русскую букву на английский аналог
                char = eng_letters[rus_letters.index(char)]
            message_idx += 1
        container_with_hidden_info.append(char)
    
    return ''.join(container_with_hidden_info)

# Чтение файлов и обработка
with open(container_file, 'r', encoding='cp1251') as container_f:
    container_text = container_f.read()

with open(secret_message_file, 'r', encoding='cp1251') as secret_f:
    secret_message = secret_f.read()

# Преобразование секретного сообщения в двоичный формат
secret_message_bin = text_to_bin(secret_message)

# Скрытие информации
container_with_hidden_info = hide_info_in_container(container_text, secret_message_bin)

# Запись результата в файл
with open(output_file, 'w', encoding='cp1251') as output_f:
    output_f.write(container_with_hidden_info)

File: task 2/hide.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <bitset>
#include <algorithm>

using namespace std;

const std::vector<wchar_t> rus_letters = {L'а', L'е', L'о', L'р', L'с', L'у', L'х', L'А', L'В', L'Е', L'К', L'О', L'Р', L'С', L'Т', L'Х'};
const std::vector<wchar_t> eng_letters = {L'a', L'e', L'o', L'p', L'c', L'y', L'x', L'A', L'B', L'E', L'K', L'O', L'P', L'C', L'T', L'X'};

std::wstring text_to_bin(const std::wstring& text) {
    std::wstring binary_str;
    for (wchar_t ch : text) {
        binary_str += std::bitset<8>(static_cast<unsigned char>(ch)).to_string();
    }
    return binary_str;
}

std::wstring hide_info_in_container(const std::wstring& container, const std::wstring& secret_message_bin) {
    std::wstring container_with_hidden_info;
    size_t message_idx = 0;

    for (wchar_t ch : container) {
        if (message_idx < secret_message_bin.length() && std::find(rus_letters.begin(), rus_letters.end(), ch) != rus_letters.end()) {
            if (secret_message_bin[message_idx] == L'1') {
                // Заменяем русскую букву на английский аналог
                ch = eng_letters[std::distance(rus_letters.begin(), std::find(rus_letters.begin(), rus_letters.end(), ch))];
            }
            message_idx++;
        }
        container_with_hidden_info += ch;
    }

    return container_with_hidden_info;
}

int main() {
    std::wstring container_file = L"container.txt";
    std::wstring secret_message_file = L"secret.txt";
    std::wstring output_file = L"code.txt";

    std::wifstream container_f(container_file);
    std::wifstream secret_f(secret_message_file);
    std::wofstream output_f(output_file);

    if (!container_f.is_open() || !secret_f.is_open() || !output_f.is_open()) {
        std::wcerr << L"Ошибка при открытии файлов!" << std::endl;
        return 1;
    }

    std::wstring container_text((std::istreambuf_iterator<wchar_t>(container_f)), std::istreambuf_iterator<wchar_t>());
    std::wstring secret_message((std::istreambuf_iterator<wchar_t>(secret_f)), std::istreambuf_iterator<wchar_t>());

    std::wstring secret_message_bin = text_to_bin(secret_message);
    std::wstring container_with_hidden_info = hide_info_in_container(container_text, secret_message_bin);

    output_f << container_with_hidden_info;

    container_f.close();
    secret_f.close();
    output_f.close();

    return 0;
}
